core/domain/classifier_services.py

```diff
@@ -0,0 +1,507 @@
+    word instance."""
+
+    _DEFAULT_ALPHA = 0.1
+    _DEFAULT_BETA = 0.001
+
+    _DEFAULT_TRAINING_ITERATIONS = 25
+    _DEFAULT_PREDICTION_ITERATIONS = 5
+
+    _DEFAULT_PREDICTION_THRESHOLD = 0.5
+
+    _DEFAULT_LABEL = '_default'
+
+    def __init__(self):
+        """ Initializes constants for the classifier. Setting a seed ensures
+
+        that results are deterministic. There is nothing special about the
+        value 4. These should not be changed unless you know what you're
+        doing."""
+        numpy.random.seed(seed=4)
+
+        self._alpha = self._DEFAULT_ALPHA
+        self._beta = self._DEFAULT_BETA
+
+        self._training_iterations = self._DEFAULT_TRAINING_ITERATIONS
+        self._prediction_iterations = self._DEFAULT_PREDICTION_ITERATIONS
+
+        self._prediction_threshold = self._DEFAULT_PREDICTION_THRESHOLD
+
+    def _get_word_id(self, word):
+        """Returns a word's id if it exists, otherwise assigns
+
+        a new id to the word and returns it."""
+        if word not in self._word_to_id:
+            self._word_to_id[word] = self._num_words
+            self._num_words += 1
+        return self._word_to_id[word]
+
+    def _get_label_id(self, label):
+        """Returns a label's id if it exists, otherwise assigns
+
+        a new id to the label and returns it."""
+        if label not in self._label_to_id:
+            self._label_to_id[label] = self._num_labels
+            self._num_labels += 1
+        return self._label_to_id[label]
+
+    def _get_label_name(self, l):
+        """Returns a label's string name given its internal id.
+
+        If the id does not have a corresponding name, an exception is
+        raised."""
+        for label_name, label_id in self._label_to_id.iteritems():
+            if label_id == l:
+                return label_name
+
+        raise Exception('Label id %d does not exist.' % l)
+
+    def _get_doc_with_label_vector(self, d):
+        """Given a doc id, return the doc and its label bit vector."""
+        return self._w_dp[d], self._b_dl[d]
+
+    def _label_set_to_vector(self, labels):
+        """Generate and return a label bit vector given a list of labels that
+        are turned on for the vector."""
+        label_vector = numpy.zeros(self._num_labels)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        label_vector[self._label_to_id[self._DEFAULT_LABEL]] = 1
+        return label_vector
+
+    def _get_label_vector(self, labels):
+        """Returns a label bit vector given a list of labels.
+
+        An empty label list is an alias for all labels being set. This is
+        useful because no old input data needs to be altered when a new label
+        is introduced to the model."""
+        if len(labels) == 0:
+            return self._label_set_to_vector([
+                label_to_id[0] for label_to_id in self._label_to_id.items()])
```

Comment:
I'm a little confused here, sorry -- isn't label_to_id an int?

Also can we talk about this a bit?